### ISSUE TITLE
Configuring Proxy (with optional credentials)

### ISSUE THREAD

**Issue by klem231188** [bytefish/FcmJava#17]
Hello.
Presently there is no way to configure a proxy in the HttpClient class.
It would be nice if we could configure a proxy (hostname + port) and optional credentials (user + password).
in a properties file.
Thanks for your library.

**Comment by bytefish** [comment]
The Apache HttpClient is constructed in the ``HttpClient`` class. I think it would be useful to add an additional method to configure the Apache ``HttpClientBuilder``, which is used for requests. But I do not have a lot of time right now and I don't know if I can implement it the weeks ahead.

**Comment by bytefish** [comment]
I have added an additional ``configure``method in the ``HttpClient`` class.

It can be used like this:

```java
// Copyright (c) Philipp Wagner. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

package de.bytefish.fcmjava.client.tests;

import de.bytefish.fcmjava.client.FcmClient;
import de.bytefish.fcmjava.client.http.HttpClient;
import de.bytefish.fcmjava.http.client.IFcmClient;
import de.bytefish.fcmjava.http.options.IFcmClientSettings;
import org.apache.http.HttpHost;
import org.apache.http.auth.AuthScope;
import org.apache.http.auth.UsernamePasswordCredentials;
import org.apache.http.impl.client.BasicCredentialsProvider;
import org.apache.http.impl.client.ProxyAuthenticationStrategy;
import org.junit.Test;

class FakeFcmClientSettings implements IFcmClientSettings {

    @Override
    public String getFcmUrl() {
        return "";
    }

    @Override
    public String getApiKey() {
        return "";
    }
}

public class HttpBuilderConfigurationTest {


    @Test
    public void testFcmClientWithProxySettings() {

        // Create Settings:
        IFcmClientSettings settings = new FakeFcmClientSettings();

        // Create the HttpClient:
        HttpClient httpClient = new HttpClient(settings);

        // And configure the HttpClient:
        httpClient.configure((httpClientBuilder -> {

            // Define the Credentials to be used:
            BasicCredentialsProvider basicCredentialsProvider = new BasicCredentialsProvider();

            // Set the Credentials (any auth scope used):
            basicCredentialsProvider.setCredentials(AuthScope.ANY, new UsernamePasswordCredentials("your_username", "your_password"));

            httpClientBuilder
                    // Set the Proxy Address:
                    .setProxy(new HttpHost("your_hostname", 1234))
                    // Set the Authentication Strategy:
                    .setProxyAuthenticationStrategy(new ProxyAuthenticationStrategy())
                    // Set the Credentials Provider we built above:
                    .setDefaultCredentialsProvider(basicCredentialsProvider);
        }));

        // Finally build the FcmClient:
        IFcmClient client = new FcmClient(settings, httpClient);
    }
}
```

**Comment by klem231188** [comment]
Thank you for your speed enhancement.
When do you plan to release it ?(and publish it on maven repository)

**Comment by bytefish** [comment]
I wanted to push a new release, but my GPG key expired. I have probably set a short expiration date, when creating the keys initially. I have extended the expiration date and pushed it to the keyserver, but the Maven is still making problems. 

I have to investigate it. This will probably take some time until the weekend, because I do not have any time left right now.

**Comment by bytefish** [comment]
I have released version 0.9 to the Central Maven Repository. It should be available in some hours, [when the Central Maven Repository synchronizes the repositories](http://stackoverflow.com/a/23236185/513875)).
